pkg/ingester/client/wrapped.go

```diff
@@ -0,0 +1,389 @@
+				return ErrIntOverflowIngester
+			}
+			if iNdEx >= l {
+				return io.ErrUnexpectedEOF
+			}
+			b := dAtA[iNdEx]
+			iNdEx++
+			wire |= uint64(b&0x7F) << shift
+			if b < 0x80 {
+				break
+			}
+		}
+		fieldNum := int32(wire >> 3)
+		wireType := int(wire & 0x7)
+		if wireType == 4 {
+			return fmt.Errorf("proto: QueryStreamResponse: wiretype end group for non-group")
+		}
+		if fieldNum <= 0 {
+			return fmt.Errorf("proto: QueryStreamResponse: illegal tag %d (wire type %d)", fieldNum, wire)
+		}
+		switch fieldNum {
+		case 1:
+			if wireType != 2 {
+				return fmt.Errorf("proto: wrong wireType = %d for field Chunkseries", wireType)
+			}
+			var msglen int
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return ErrIntOverflowIngester
+				}
+				if iNdEx >= l {
+					return io.ErrUnexpectedEOF
+				}
+				b := dAtA[iNdEx]
+				iNdEx++
+				msglen |= int(b&0x7F) << shift
+				if b < 0x80 {
+					break
+				}
+			}
+			if msglen < 0 {
+				return ErrInvalidLengthIngester
+			}
+			postIndex := iNdEx + msglen
+			if postIndex < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			// Use wrapped TimeSeriesChunk type to allow chunk buffer reuse.
+			ts := WrappedTimeSeriesChunk{TimeSeriesChunk: &TimeSeriesChunk{}}
+			if err := ts.Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+				return err
+			}
+			m.Chunkseries = append(m.Chunkseries, *ts.TimeSeriesChunk)
+			iNdEx = postIndex
+		case 2:
+			if wireType != 2 {
+				return fmt.Errorf("proto: wrong wireType = %d for field Timeseries", wireType)
+			}
+			var msglen int
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return ErrIntOverflowIngester
+				}
+				if iNdEx >= l {
+					return io.ErrUnexpectedEOF
+				}
+				b := dAtA[iNdEx]
+				iNdEx++
+				msglen |= int(b&0x7F) << shift
+				if b < 0x80 {
+					break
+				}
+			}
+			if msglen < 0 {
+				return ErrInvalidLengthIngester
+			}
+			postIndex := iNdEx + msglen
+			if postIndex < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			m.Timeseries = append(m.Timeseries, mimirpb.TimeSeries{})
+			if err := m.Timeseries[len(m.Timeseries)-1].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+				return err
+			}
+			iNdEx = postIndex
+		default:
+			iNdEx = preIndex
+			skippy, err := skipIngester(dAtA[iNdEx:])
+			if err != nil {
+				return err
+			}
+			if skippy < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if (iNdEx + skippy) < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if (iNdEx + skippy) > l {
+				return io.ErrUnexpectedEOF
+			}
+			iNdEx += skippy
+		}
+	}
+
+	if iNdEx > l {
+		return io.ErrUnexpectedEOF
+	}
+	return nil
+}
+
+type WrappedTimeSeriesChunk struct {
+	*TimeSeriesChunk
+}
+
+func (m *WrappedTimeSeriesChunk) Reset() {
+	*m = WrappedTimeSeriesChunk{&TimeSeriesChunk{}}
+}
+
+func (m *WrappedTimeSeriesChunk) Unmarshal(dAtA []byte) error {
```

Comment:
This seems to be autogenerated code, but it's unclear where it comes from or how to rebuild it again.

IMO having a code like this isn't worth it unless there's a HUGE performance improvement.

Reply:
> IMO having a code like this isn't worth it unless there's a HUGE performance improvement.

My first reaction was the same, but 1) we already have similar code (https://github.com/grafana/mimir/blob/062807abb982320d603fe7661fa6f2e1e581f7bc/pkg/mimirpb/timeseries.go#L131) and 2) benchmarks here don't look bad (although they need proper review too).

Reply:
I agree that the benchmarks look good, then if we're keeping this code, we should document:
- What's the purpose of each of the new structs.
- How did we generate the unreadable Unmarshal() method (see the comment on the one @pstibrany linked)
- What are the expectations from the user of this entity (chunks have to be put back into the pool).

Reply:
Agree with @colega - I'd also like to see documentation of the origin of auto-generated code. I could only deduce this is possibly (partially?) auto-generated. E.g. which portions of this code is auto-generated, and which are not?